core/src/main/scala/com/metamx/tranquility/finagle/DruidTaskResolver.scala

```diff
@@ -0,0 +1,144 @@
+  )
+
+  private val started      = new AtomicBoolean
+  private val taskNumberer = new AtomicLong
+
+  private val lock     = new AnyRef
+  private var nextPoll = 0L
+  private var tasks    = Map[Long, (String, AtomicReference[Addr], Updatable[Addr])]()
+
+  private def ensureStarted() {
+    if (started.compareAndSet(false, true)) {
+      log.info(s"Starting poll thread for indexService[$indexService], pollPeriod[$pollPeriod].")
+      pollExecutor.submit(
+        new Runnable
+        {
+          override def run(): Unit = {
+            while (!Thread.currentThread().isInterrupted) {
+              try {
+                val tasksSnapshot: Map[Long, (String, AtomicReference[Addr], Updatable[Addr])] = lock.synchronized {
+                  while (timekeeper.now.millis < nextPoll) {
+                    val waitMillis = nextPoll - timekeeper.now.millis
+                    if (waitMillis > 0) {
+                      lock.wait(waitMillis)
+                    }
+                  }
+                  tasks
+                }
+
+                val now = timekeeper.now
+                val baseWait = now.plus(pollPeriod).millis - now.millis
+                val fuzzyWait = (math.max(1 + 0.25 * Random.nextGaussian, 0) * baseWait).toLong
+                nextPoll = now.millis + fuzzyWait
+
+                val running = Await.result(indexService.runningTasks())
+
+                log.info(s"Polled[$indexServiceKey] and found ${running.size} tasks " +
+                  s"(${running.count(_._2.isBound)} bound).")
+
+                for ((taskNumber, (taskId, addr, updatable)) <- tasksSnapshot) {
+                  val runningHostPort = running.get(taskId)
+                  val runningAddr: Addr = runningHostPort.map(_.toAddr).getOrElse(Addr.Neg)
+                  if (addr.get() != runningAddr) {
+                    log.info(s"Updating location for task[$taskId] (#$taskNumber) to[$runningAddr].")
+                    addr.set(runningAddr)
+                    updatable.update(runningAddr)
+                  }
+                }
+              }
+              catch {
+                case NonFatal(e) =>
+                  log.warn(e, s"Poll failed, trying again at[${new DateTime(nextPoll)}].")
+              }
+            }
+          }
+        }
+      )
+    }
+  }
+
+  override val scheme = DruidTaskResolver.mkscheme(indexService.key)
+
+  override def bind(taskId: String) = {
+    val taskNumber = taskNumberer.incrementAndGet()
+
+    Var.async[Addr](Addr.Pending) {
+      updatable =>
+        log.info(s"Started monitoring task[$taskId] (#$taskNumber).")
+
+        lock.synchronized {
+          nextPoll = 0L
+          tasks = tasks + (taskNumber ->(taskId, new AtomicReference[Addr], updatable))
+          lock.notifyAll()
+        }
+
+        ensureStarted()
+        new Closable
+        {
+          override def close(deadline: Time): Future[Unit] = {
+            lock.synchronized {
```

Comment:
since deadline is only advisory this can block for as long as it needs to I suppose.